services/supervisor/src/transactions.rs

```diff
@@ -413,6 +447,150 @@ impl SupervisorInterface<ExecutionContext<'_>> for Supervisor {
+                // Migration failed even before start, softly mark it as failed.
+                let initiate_rollback = false;
+                return self.fail_migration(context, request, error, initiate_rollback);
+            }
+
+            // Migration started. Check if migration is fast-forward.
+            let instance = get_instance_by_name(&context, request.service.as_ref())
+                .expect("BUG: Instance disappeared");
+            if state.version != *instance.data_version() {
+                // Migration is fast-forward, complete it immediately.
+                // No agreement needed, since nodes which will behave differently will obtain
+                // different blockchain state hash and will be excluded from consensus.
+                log::trace!("Applied fast-forward migration with request {:?}", request);
+
+                let mut schema = SchemaImpl::new(context.service_data());
+
+                // Update the state of a migration.
+                let mut state = schema
+                    .migration_states
+                    .get(&request)
+                    .expect("BUG: Migration succeed, but does not have a stored state");
+                state.update(AsyncEventState::Succeed, instance.data_version().clone());
+                schema.migration_states.put(&request, state);
+
+                // Remove the migration from the list of pending.
+                schema.pending_migrations.remove(&request);
+            }
+        }
+        Ok(())
+    }
+
+    fn report_migration_result(
+        &self,
+        context: ExecutionContext<'_>,
+        result: MigrationResult,
+    ) -> Self::Output {
+        // Verifies that transaction author is validator.
+        let author = get_validator(&context)?;
+
+        result
+            .request
+            .new_artifact
+            .validate()
+            .map_err(|e| ArtifactError::InvalidArtifactId.with_description(e))?;
+
+        let core_schema = context.data().for_core();
+        let current_height = core_schema.height();
+
+        let schema = SchemaImpl::new(context.service_data());
+
+        // Verify that this migration is registered.
+        schema
+            .migration_states
+            .get(&result.request)
+            .ok_or(MigrationError::MigrationRequestNotRegistered)?;
+
+        // Check if migration already failed.
+        if schema
+            .migration_states
+            .get(&result.request)
+            .map(|state| state.is_failed())
+            .unwrap_or_default()
+        {
+            // This migration is already resulted in failure, no further
+            // processing needed.
+            return Ok(());
+        }
+
+        // Verify that we didn't reach deadline height.
+        if result.request.deadline_height < current_height {
+            return Err(SupervisorCommonError::DeadlineExceeded.into());
+        }
+
+        drop(schema);
+
+        match result.result.0 {
```

Comment:
Nit: `result.result` duplication. As a suggestion, could it make sense to name the field `status` (given its type - `MigrationStatus`)?